File: reverse-linked-list/Solution206.java
// https://leetcode.com/problems/reverse-linked-list/

import java.util.Stack;

class Solution206 {

    // T: O(n), S: O(1)
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode prev = null, curr = head;
        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    // T: O(n), S: O(n)
    public ListNode reverseListStack(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        var st = new Stack<ListNode>();
        var curr = head;
        while (curr != null) {
            st.push(curr);
            curr = curr.next;
        }
        var newHead = st.pop();
        curr = newHead;
        while (!st.isEmpty()) {
            curr.next = st.pop();
            curr = curr.next;
        }
        curr.next = null;
        return newHead;
    }

    // T: O(n), S: O(n)
    public ListNode reverseListRec(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode reversedListHead = reverseListRec(head.next);
        head.next.next = head;
        head.next = null;
        return reversedListHead;
    }

    public static void main(String[] args) {
        var sol = new Solution206();
        var list1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null)))));
        System.out.println(list1); // 1 -> 2 -> 3 -> 4 -> 5
        System.out.println(sol.reverseList(list1)); // 5 -> 4 -> 3 -> 2 -> 1

        var list2 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null))));
        System.out.println(list2); // 1 -> 2 -> 3 -> 4
        System.out.println(sol.reverseListStack(list2)); // 4 -> 3 -> 2 -> 1

        var list3 = new ListNode(1, new ListNode(2, new ListNode(3, null)));
        System.out.println(list3); // 1 -> 2 -> 3
        System.out.println(sol.reverseListRec(list3)); // 3 -> 2 -> 1
    }

}

File: reverse-linked-list/solution206.go
// https://leetcode.com/problems/reverse-linked-list/

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// T: O(n), S: O(1)
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var prev *ListNode
	curr := head
	for curr != nil {
		next := curr.Next
		curr.Next = prev
		prev = curr
		curr = next
	}
	return prev
}

// T: O(n), S: O(n)
func reverseListStack(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	stack := []*ListNode{}
	curr := head
	for curr != nil {
		stack = append(stack, curr)
		curr = curr.Next
	}
	newHead := stack[len(stack)-1]
	stack = stack[:len(stack)-1]
	curr = newHead
	for len(stack) > 0 {
		curr.Next = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		curr = curr.Next
	}
	curr.Next = nil
	return newHead
}

// T: O(n), S: O(n)
func reverseListRec(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	reversedListHead := reverseListRec(head.Next)
	head.Next.Next = head
	head.Next = nil
	return reversedListHead
}

func printList(head *ListNode) {
	curr := head
	for curr != nil {
		fmt.Print(curr.Val)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func main() {
	list1 := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{5, nil}}}}}
	printList(list1)              // 1 -> 2 -> 3 -> 4 -> 5
	printList(reverseList(list1)) // 5 -> 4 -> 3 -> 2 -> 1

	list2 := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, nil}}}}
	printList(list2)                   // 1 -> 2 -> 3 -> 4
	printList(reverseListStack(list2)) // 4 -> 3 -> 2 -> 1

	list3 := &ListNode{1, &ListNode{2, &ListNode{3, nil}}}
	printList(list3)                 // 1 -> 2 -> 3
	printList(reverseListRec(list3)) // 3 -> 2 -> 1
}
